app/src/main/java/org/schabi/newpipe/util/SparseItemUtil.java

```diff
@@ -0,0 +1,126 @@
+     */
+    public static void fetchItemInfoIfSparse(@NonNull final Context context,
+                                             @NonNull final StreamInfoItem item,
+                                             @NonNull final Consumer<SinglePlayQueue> callback) {
+        if ((!(item.getStreamType() == LIVE_STREAM || item.getStreamType() == AUDIO_LIVE_STREAM)
+                && item.getDuration() < 0) || isNullOrEmpty(item.getUploaderUrl())) {
+            fetchStreamInfoAndSaveToDatabase(context, item.getServiceId(), item.getUrl(),
+                    streamInfo -> callback.accept(new SinglePlayQueue(streamInfo)));
+        } else {
+            // all info is already there, no need to fetch
+            callback.accept(new SinglePlayQueue(item));
+        }
+    }
+
+    /**
+     * Use this to certainly obtain an uploader url when the stream info item or play queue item you
+     * are handling might not have the uploader url (e.g. because it was fetched with {@link
+     * org.schabi.newpipe.extractor.feed.FeedExtractor}). A toast is shown if loading details is
+     * required.
+     *
+     * @param context the Android context
+     * @param serviceId the serviceId of the item
+     * @param url the item url
+     * @param uploaderUrl the uploaderUrl of the item; if null or empty will be fetched
+     * @param callback the callback called with either the original uploaderUrl, if it was a valid
+     *                 url, otherwise with the uploader url obtained by fetching the {@link
+     *                 org.schabi.newpipe.extractor.stream.StreamInfo} corresponding to the item
+     */
+    public static void fetchUploaderUrlIfSparse(@NonNull final Context context,
+                                                final int serviceId,
+                                                @NonNull final String url,
+                                                @Nullable final String uploaderUrl,
+                                                @NonNull final Consumer<String> callback) {
+        if (isNullOrEmpty(uploaderUrl)) {
+            fetchStreamInfoAndSaveToDatabase(context, serviceId, url,
+                    streamInfo -> callback.accept(streamInfo.getUploaderUrl()));
+        } else {
+            callback.accept(uploaderUrl);
+        }
+    }
+
+    /**
+     * Loads the stream info corresponding to the given data on an I/O thread, stores the result in
+     * the database and calls the callback on the main thread with the result. A toast will be shown
+     * to the user about loading stream details, so this needs to be called on the main thread.
+     *
+     * @param context the Android context
+     * @param serviceId the service id of the stream to load
+     * @param url the url of the stream to load
+     * @param callback the callback to call with the result
+     */
+    private static void fetchStreamInfoAndSaveToDatabase(final Context context,
```

Comment:
The context should be annotated with `@NonNull`, because it must not be null to show a toast (and maybe in more places).

Reply:
Done and did not find other places